lightningd/peer_control.c

```diff
@@ -942,6 +944,75 @@ static int peer_got_funding_locked(struct peer *peer, const u8 *msg)
 	return 0;
 }
 
+static u8 *p2wpkh_for_keyidx(const tal_t *ctx, struct lightningd *ld, u64 keyidx)
+{
+	struct pubkey shutdownkey;
+
+	if (!bip32_pubkey(ld->bip32_base, &shutdownkey, keyidx))
+		return NULL;
+
+	return scriptpubkey_p2wpkh(ctx, &shutdownkey);
+}
+
+static int peer_got_shutdown(struct peer *peer, const u8 *msg)
+{
+	u8 *scriptpubkey;
+
+	if (!fromwire_channel_got_shutdown(peer, msg, NULL, &scriptpubkey)) {
+		log_broken(peer->log, "bad channel_got_funding_locked %s",
+			   tal_hex(peer, msg));
+		return -1;
+	}
+
+	/* FIXME: Add to spec that we must allow repeated shutdown! */
+	peer->remote_shutdown_scriptpubkey
+		= tal_free(peer->remote_shutdown_scriptpubkey);
+	peer->remote_shutdown_scriptpubkey = scriptpubkey;
+	/* FIXME: Save to db */
+
+	if (peer->local_shutdown_idx == -1) {
+		u8 *scriptpubkey;
+
+		peer->local_shutdown_idx = wallet_get_newindex(peer->ld);
+		if (peer->local_shutdown_idx == -1) {
+			peer_internal_error(peer,
+					    "Can't get local shutdown index");
+			return -1;
+		}
+		/* FIXME: Save to db */
```

Comment:
We need to save that it's for this peer: if peer restarts, we need to tell it the shutdown scriptpubkey.